docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+and locking from the IPC transport: If all requests and responses are routed
+through one connection, the server can explicitly maintain priority queues for
+requests and responses, and make less-conservative decisions about when locks
+are (or aren't) required to synchronize state access. With independent queues,
+the server must lock conservatively, and no optimistic scheduling is practical.
+
+This would be a tedious implementation change, but should be achievable without
+breaking any of the existing interfaces. More importantly, it could potentially
+address a lot of difficult concurrency and performance problems we currently
+see anecdotally but have difficultly isolating because of how intertwined these
+separate message streams are at runtime.
+
+TODO: Impact of ABCI++ for this topic?
+
+### Context: RPC Issues
+
+The RPC system serves several masters, and has a complex surface area. I
+believe there are some improvements that can be exposed by separating some of
+these concerns.
+
+For light clients, some work is already underway to move toward using P2P
+message passing rather than RPC to synchronize light client state with the rest
+of the network.  Because of the way light client synchronization interacts with
+event logging, this is not as simple as just swapping out the transport: More
+details of the event system will need to be made visible via P2P, and that in
+turn has some implications for fault tolerance (in particular, I think the
+ability of a badly-behaved client to DDoS the network without punishment is a
+concern). However, once this work is done, it should in principle be possible
+to remove the dependency of light clients on the RPC service.
+
+For bootstrapping and operations, the RPC mechanism still makes sense, but can
+be further simplified. Here are some specific proposals:
+
+- Remove the HTTP GET interface entirely.
+- Simplify JSON-RPC plumbing to remove unnecessary reflection and wrapping.
+- Remove the gRPC interface (this is already planned for v0.36).
+- Remove the websocket interface (possibly depends on light client changes).
+
+These changes would preserve the ability of operators to issue queries with
+curl (but would require using JSON-RPC instead of URI parameters). That would
+be a little less user-friendly, but for a use case that should not be that
+prevalent.
+
+These changes would also preserve compatibility with existing JSON-RPC based
+code paths like the `tendermint` CLI and the light client (even ahead of
+further work to remove that dependency).
+
+**Design goal:** An operator should be able to disable non-local access to the
+RPC server on any node in the network without impairing the ability of the
+network to function for service of state replication, including light clients.
+
+**Design principle:** All communication required to implement and monitor the
+consensus network should use P2P, including the various synchronizations.
```

Comment:
The way I've viewed it (and YMMV) is that any inter node communication (including the light client) should use p2p. Any communication from external clients to the network of nodes should use RPC (either for sending transactions or querying)

Reply:
> Any communication from external clients to the network of nodes should use RPC (either for sending transactions or querying)

I agree about internode communication. On this second point, though, I think it depends what you mean by RPC: I'd argue a client who wants to put transactions into the network or query application status should be calling the _application_, not the consensus node. At least for applications built on the SDK, I think that already works (via gRPC).  Queries for _consensus_ status do make sense via the Tendermint RPC interface, though.

Reply:
Currently the only way to get a transaction into the network is via the Tendermint RPC. If the CosmosSDK enables this via their gRPC then I assume it's calling the Tendermint endpoint underneath.

Reply:
> Currently the only way to get a transaction into the network is via the Tendermint RPC. If the CosmosSDK enables this via their gRPC then I assume it's calling the Tendermint endpoint underneath.

Right. In order to _fully_ implement some of the proposals I've sketched here, we would need to address that.

Even if we can't do that right away, though, we can still make progress: The application and the consensus node are either linked together (via the SDK) or in adjacent processes (and hence share a network segment). So even though the application now needs to call the RPC service to get work done, we could shut down remote access to Tendermint RPC without breaking the network. 

I think it's _also_ worth breaking that dependency, but doing that is of lower priority IMO.